### DISCUSSION TITLE
Implementing Terragrunt on existing Terraform codebase

### DISCUSSION THREAD

**Discussion by lpossamai** [gruntwork-io/knowledge-base#740] [How do I]
Hi everyone!

I have a pure Terraform repository with many modules (each on their own state file). I'm planning to implement Terragrunt. At the moment, the repository is using TF workspaces to deploy to different accounts (dev, test, staging, etc).

Was wondering if anyone has ever done a migration like that and could share some suggestions/thoughts?

Thanks in advance!

---

<ins datetime="2023-06-15T23:32:28Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110259">Tracked in ticket #110259</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
exists a couple of articles and example repositories to keep configuration dry and deploy for multiple environments:

- https://github.com/gruntwork-io/terragrunt-infrastructure-live-example
- https://github.com/gruntwork-io/terragrunt-infrastructure-modules-example
- https://terragrunt.gruntwork.io/docs/features/keep-your-terraform-code-dry/
- https://blog.gruntwork.io/how-to-manage-multiple-environments-with-terraform-using-terragrunt-2c3e32fc60a8
- https://blog.gruntwork.io/how-to-manage-multiple-environments-with-terraform-using-workspaces-98680d89a03e
- https://blog.gruntwork.io/how-to-manage-multiple-environments-with-terraform-32c7bc5d692

**Reply by lpossamai** [reply]
Yeah.. I was more looking for advises on the migration itself... I'm already using those examples above to structure my repo.

**Comment by pras111gw** [comment]
This is an interesting use case, and we are glad you are considering terragrunt. In addition to @denis256 notes, I think migration from workspaces' state files to under terragrunt management should go smoothly. 

1. First, back up your state files to a new bucket just in case something goes wrong. 
2. Ensure s3 bucket versioning is turned on.
3. Inventory your current state files. Get their S3 bucket names, dynamodb lock tables etc.
4. Reorganize your code using the terragrunt-infra-live example that @denis256 shared. See attached sample for a terragrunt-managed repo.
5. Overwrite terragrunt backend settings with your own ones from workspaces. 

![image](https://github.com/gruntwork-io/knowledge-base/assets/116776318/091b33e4-415a-40ac-acce-e284006e5749)

**Comment by lpossamai** [comment]
This is interesting https://devpress.csdn.net/cicd/62ec396f89d9027116a10733.html

**Comment by lpossamai** [comment]
So, looks like there is no easy way to implement Terragrunt on existing (pure) Terraform repository.

You'll need to create the Terragrunt configuration files and import one by one using `terragrunt import`. Sure, you can develop a script that does that.
